### PR TITLE
#347 Allow custom layouts in Button with labelWidget property

### PR DESCRIPTION
**Pull request by evshi**

### CODE REVIEW

File: lib/widget/button.dart [hunk 1]
```diff
@@ -29,6 +31,7 @@ class Button extends StatefulWidget with Invokable, HasController<ButtonControll
   Map<String, Function> setters() {
     return {
       'label': (value) => _controller.label = Utils.getString(value, fallback: ''),
+      'labelWidget': (value) => _controller.labelWidget = value,
```
Comment (vusters) [inline lib/widget/button.dart:34]: we do this everywhere for widget but something to think about: we have to assume what the user enters here will resolve to a valid widget later (or else we show 'invalid widget' text). Probably need a `Utils.getWidget(value)` here where we check against the list of valid widgets, plus any user-defined custom widget.

**Review by vusters — APPROVED** [review]

**Comment by vusters** [comment]
You can add the schema and add to Kitchen sink example now please https://studio.ensembleui.com/app/e24402cb-75e2-404c-866c-29e6c3dd7992/screen/09c1087b-f9ee-4a8c-9286-e0e881184c07

File: lib/widget/button.dart (hunk 1 above)
Comment (evshi) [reply lib/widget/button.dart:34]: The check makes sense but how could we handle it differently if it doesn't resolve to a valid widget? For example, fail silently or return a default widget?

Comment (evshi) [reply lib/widget/button.dart:34]: Got it. Shall I create a new issue for the widget util validation or is there one already?

Comment (vusters) [reply lib/widget/button.dart:34]: https://github.com/orgs/EnsembleUI/projects/2/views/1?pane=issue&itemId=23052103

File: assets/schema/ensemble_schema.json [hunk 2]
```diff
@@ -814,12 +814,15 @@
...
       "type": "object",
-      "required": ["label"],
       "properties": {
         "label": {
           "type": "string",
           "description": "The button label"
         },
+        "labelWidget": {
+          "$ref": "#/$defs/Row-payload",
+          "description": "A row widget containing the button contents. For example, a Text and an Icon"
```
Comment (vusters) [inline assets/schema/ensemble_schema.json:824]: Use this: "the custom widget to display in place of the label."

**Review by vusters — APPROVED** [review]
Change the widget reference and you should be good to merge

File: lib/widget/button.dart [hunk 3]
```diff
@@ -82,20 +90,39 @@ class ButtonState extends WidgetState<Button> {
...
+      endingIcon = scopeManager?.buildWidgetFromDefinition(widget._controller.endingIcon);
+    }
+    List<Widget> labelParts = [Text(Utils.translate(widget._controller.label ?? '', context))];
+    if (startingIcon != null) {
+      labelParts.insertAll(0, [startingIcon]);
+    }
+    if (endingIcon != null) {
+      labelParts.addAll([endingIcon]);
+    }
+    Widget labelLayout = Row(
```
Comment (evshi) [inline lib/widget/button.dart:111]: @vusters it seems like Row is required to render icons after the button and have the ability to have both icons. Another benefit is Row automatically responds to the device text direction, i.e. RTL. Let me know if this is concerning to you or you think we should move in a different direction.

File: lib/widget/button.dart [hunk 4]
```diff
@@ -89,25 +92,37 @@ class ButtonState extends WidgetState<Button> {
...
-    Widget label;
-    if (scopeManager != null && widget._controller.labelWidget != null) {
-      label = scopeManager.buildWidgetWithScopeFromDefinition(widget._controller.labelWidget);
-    } else {
-      label = Text(
-          Utils.translate(widget._controller.label ?? '', context));
+    Widget? startingIcon;
+    Widget? endingIcon;
+    if (widget._controller.startingIcon != null) {
+      startingIcon = scopeManager?.buildWidgetFromDefinition(
```
Comment (vusters) [inline lib/widget/button.dart:98]: use Icon.fromModel(IconModel)

File: lib/widget/button.dart (hunk 3 above)
Comment (vusters) [reply lib/widget/button.dart:111]: This is fine.

File: lib/widget/button.dart [hunk 5]
```diff
@@ -89,25 +92,37 @@ class ButtonState extends WidgetState<Button> {
...
+          widget._controller.startingIcon);
+    }
+    if (widget._controller.endingIcon != null) {
+      endingIcon = scopeManager?.buildWidgetFromDefinition(widget._controller.endingIcon);
+    }
+    List<Widget> labelParts = [Text(Utils.translate(widget._controller.label ?? '', context))];
```
Comment (vusters) [inline lib/widget/button.dart:104]: User may want to control the gap between the icons and label. Add an `iconGap` attribute? by default it should be 0.

File: lib/widget/button.dart [hunk 6]
```diff
@@ -89,25 +92,37 @@ class ButtonState extends WidgetState<Button> {
...
+    if (startingIcon != null) {
+      labelParts.insertAll(0, [startingIcon]);
```
Comment (vusters) [inline lib/widget/button.dart:106]: why not just use `add(endingIcon)` and `insert(0, startingIcon)`?

Comment (evshi) [reply lib/widget/button.dart:106]: This is a holdover from when I was going to add gaps in between the icon and label.

File: lib/widget/button.dart (hunk 5 above)
Comment (evshi) [reply lib/widget/button.dart:104]: This is easy enough to add, but I think this is a slippery slope where we might consider going back to a custom label widget Row so they can control things like spacing without needing to pixel push. Do you feel strongly we should add this now or can we defer this until we get some more user feedback?

**Review by vusters — APPROVED** [review]

**Pull request merged**